apiserver/httpcontext.go

```diff
@@ -104,9 +104,51 @@ func checkPermissions(tag names.Tag, acceptFunc common.GetAuthFunc) (bool, error
 	return false, errors.NotValidf("tag kind %v", tag.Kind())
 }
 
-// stateForRequestAuthenticatedUser is like stateForRequestAuthenticated
+// stateForMigration asserts that the incoming connection is from a user that
+// has admin permissions on the controller model. The method also gets the
+// model uuid for the model being migrated from a request header, and returns
+// the state instance for that model.
+func (ctxt *httpContext) stateForMigration(r *http.Request) (*state.State, error) {
+	st, user, err := ctxt.stateAndEntityForRequestAuthenticatedUser(r)
+	if err != nil {
+		return nil, err
+	}
+	if !st.IsController() {
+		return nil, errors.BadRequestf("model is not controller model")
+	}
+	admin, err := st.IsControllerAdmin(user.Tag().(names.UserTag))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !admin {
+		return nil, errors.Unauthorizedf("not a controller admin")
+	}
+
+	modelUUID, err := validateModelUUID(validateArgs{
+		statePool: ctxt.srv.statePool,
+		modelUUID: r.Header.Get("Migration-Model-UUID"),
+		strict:    true,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	migrationSt, err := ctxt.srv.statePool.Get(modelUUID)
```

Comment:
Hmm - actually, having errors be possible after the statePool.Get might mean that the state is leaked, because the calling code won't `defer h.ctxt.release(st)` if `err != nil`. And it couldn't because this code returns nil for the *State on error.

This will either need to release the state back to the pool on an error, or it can do the model checks using the pool's system state and then do the statePool.Get as the last thing before returning it.

Or you could defer a func that closes over named return values and releases st if the returned err != nil, but that's a bit silly.